### PR TITLE
JAMES-2720 provision of dockerized LinShare containers

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

File: third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        network = Network.newNetwork();
+        linshareDatabase = createDockerDatabase();
+        linshareMongodb = createDockerMongodb();
+        linshareLdap = createDockerLdap();
+        linshareSmtp = createDockerSmtp();
+        linshareBackend = createDockerBackend();
+        linshareDatabaseInit = createDockerDatabaseInit();
+    }
+
+    public void start() {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java]: Nope -> order is important, and staging would add some unecessary complexity

File: third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        linshareDatabase.start();
+        linshareMongodb.start();
+        linshareLdap.start();
+        linshareSmtp.start();
+        linshareBackend.start();
+        linshareDatabaseInit.start();
+    }
+
+    public void stop() {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java]: Idem

File: third-party/linshare/src/test/resources/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,77 @@
+# Docker compose file for a full-featured Linshare architecture
```
Comment (chibenwa) [inline third-party/linshare/src/test/resources/docker-compose.yml]: Please add the following disclaimer

```
# This file is not used by our Java code but Linshare class directly  implement the behaviours defined here
```

File: third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Arsnael) [reply third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java]: knowing that some containers need others started first to be able to start themselves then I would say no?

File: third-party/linshare/src/test/java/org/apache/james/linshare/LDAPConfigurationPerformer.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+import static io.restassured.config.RestAssuredConfig.newConfig;
+
+import java.nio.charset.StandardCharsets;
+
+import io.restassured.authentication.BasicAuthScheme;
+import io.restassured.builder.RequestSpecBuilder;
+import io.restassured.http.ContentType;
+import io.restassured.specification.RequestSpecification;
+
+class LDAPConfigurationPerformer {
```
Comment (trantienduchn) [inline third-party/linshare/src/test/java/org/apache/james/linshare/LDAPConfigurationPerformer.java]: Just curious, I don't see the block of code that user0, user1,... are provisioned in LDAP

File: third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    private GenericContainer createDockerSmtp() {
+        return new GenericContainer<>("linagora/opensmtpd")
+            .withNetworkAliases("smtp", "linshare_smtp")
+            .withClasspathResourceMapping("./conf/smtpd.conf",
+                "/etc/smtpd/smtpd.conf",
+                BindMode.READ_ONLY)
+            .withNetwork(network);
+    }
+
+    private GenericContainer createDockerDatabaseInit() {
```
Comment (trantienduchn) [inline third-party/linshare/src/test/java/org/apache/james/linshare/Linshare.java]: I think you don't need this method

File: third-party/linshare/src/test/java/org/apache/james/linshare/LDAPConfigurationPerformer.java (hunk 4 above)
Comment (chibenwa) [reply third-party/linshare/src/test/java/org/apache/james/linshare/LDAPConfigurationPerformer.java]: Linshare directly calls the LDAP to resolve them!

**Comment by chibenwa** [comment]
```
[109ca2f6e10d1f671df21f37b7bfc2e239c0ba9a] {"message":"OCI runtime create failed: container_linux.go:348: starting container process caused \"process_linux.go:402: container init caused \\\"rootfs_linux.go:58: mounting \\\\\\\"/james-project/third-party/linshare/target/test-classes/conf/smtpd.conf\\\\\\\" to rootfs \\\\\\\"/var/lib/docker/overlay2/b6bf90f31775e95725b78cb50b7fe87b7b019de4862d3e613255564436754983/merged\\\\\\\" at \\\\\\\"/var/lib/docker/overlay2/b6bf90f31775e95725b78cb50b7fe87b7b019de4862d3e613255564436754983/merged/etc/smtpd/smtpd.conf\\\\\\\" caused \\\\\\\"not a directory\\\\\\\"\\\"\": unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type"}
```

**of course!!!**

Alternative: on the fly dockerfiles....

Stay tuned

**Comment by chibenwa** [comment]
```
[e84aaaf2819fa65c7c23424c4d78e2bc18d6cca5] [INFO] apache-james-linshare 3.4.0-SNAPSHOT ............... SUCCESS [01:19 min]
```


WOOOOO

**Comment by chibenwa** [comment]
Can be considered :green_apple:

File: third-party/linshare/src/test/java/org/apache/james/linshare/LDAPConfigurationPerformer.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        RequestSpecification specification = new RequestSpecBuilder()
+            .setContentType(ContentType.JSON)
+            .setAccept(ContentType.JSON)
+            .setConfig(newConfig().encoderConfig(encoderConfig().defaultContentCharset(StandardCharsets.UTF_8)))
+            .setBaseUri("http://" + linshare.getIp())
+            .setPort(linshare.getPort())
+            .setAuth(basicAuthScheme)
+            .build();
+
+        String ldapId = given(specification)
```
Comment (rouazana) [inline third-party/linshare/src/test/java/org/apache/james/linshare/LDAPConfigurationPerformer.java]: s/given/with/ everywhere in this file?

Comment (chibenwa) [reply third-party/linshare/src/test/java/org/apache/james/linshare/LDAPConfigurationPerformer.java]: Can not write `with(specification)`.

Gonna try `specification.with()`

Comment (chibenwa) [reply third-party/linshare/src/test/java/org/apache/james/linshare/LDAPConfigurationPerformer.java]: `java.lang.NullPointerException: Cannot get property 'assertionClosure' on null object`

Won't change!

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**